privacy-dashboard/privacy-dashboard-impl/src/main/java/com/duckduckgo/privacy/dashboard/impl/ToggleReportsDataStore.kt

```diff
@@ -0,0 +1,243 @@
+import com.duckduckgo.privacy.dashboard.impl.SharedPreferencesToggleReportsDataStore.Keys.TOGGLE_REPORTS_PROMPTS_DISMISSED
+import com.duckduckgo.privacy.dashboard.impl.SharedPreferencesToggleReportsDataStore.Keys.TOGGLE_REPORTS_PROMPT_INTERVAL
+import com.duckduckgo.privacy.dashboard.impl.SharedPreferencesToggleReportsDataStore.Keys.TOGGLE_REPORTS_PROMPT_LIMIT_LOGIC_ENABLED
+import com.duckduckgo.privacy.dashboard.impl.SharedPreferencesToggleReportsDataStore.Keys.TOGGLE_REPORTS_RC
+import com.duckduckgo.privacy.dashboard.impl.SharedPreferencesToggleReportsDataStore.Keys.TOGGLE_REPORTS_SENT
+import com.squareup.anvil.annotations.ContributesBinding
+import dagger.SingleInstanceIn
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.stateIn
+import timber.log.Timber
+
+interface ToggleReportsDataStore {
+    fun getToggleReportsRemoteConfigJson(): String
+
+    suspend fun setToggleReportsRemoteConfigJson(value: String)
+
+    suspend fun storeDismissLogicEnabled(dismissLogicEnabled: Boolean)
+
+    suspend fun storePromptLimitLogicEnabled(promptLimitLogicEnabled: Boolean)
+
+    suspend fun storeDismissInterval(dismissInterval: Int)
+
+    suspend fun storePromptInterval(promptInterval: Int)
+
+    suspend fun storeMaxPromptCount(maxPromptCount: Int)
+
+    suspend fun insertTogglePromptDismiss()
+
+    suspend fun insertTogglePromptSend()
+
+    fun canPrompt(): Boolean
+}
+
+@ContributesBinding(AppScope::class)
+@SingleInstanceIn(AppScope::class)
+class SharedPreferencesToggleReportsDataStore @Inject constructor(
+    @ToggleReports private val store: DataStore<Preferences>,
+    @AppCoroutineScope private val appCoroutineScope: CoroutineScope,
+) : ToggleReportsDataStore {
+
+    private object Keys {
+        val TOGGLE_REPORTS_RC = stringPreferencesKey(name = "TOGGLE_REPORTS_RC")
+        val TOGGLE_REPORTS_PROMPT_LIMIT_LOGIC_ENABLED = booleanPreferencesKey(name = "TOGGLE_REPORTS_PROMPT_LIMIT_LOGIC_ENABLED")
+        val TOGGLE_REPORTS_DISMISS_LOGIC_ENABLED = booleanPreferencesKey(name = "TOGGLE_REPORTS_DISMISS_LOGIC_ENABLED")
+        val TOGGLE_REPORTS_SENT = stringSetPreferencesKey(name = "TOGGLE_REPORTS_SENT")
+        val TOGGLE_REPORTS_PROMPTS_DISMISSED = stringSetPreferencesKey(name = "TOGGLE_REPORTS_PROMPTS_DISMISSED")
+        val TOGGLE_REPORTS_DISMISS_INTERVAL = intPreferencesKey(name = "TOGGLE_REPORTS_DISMISS_INTERVAL")
+        val TOGGLE_REPORTS_PROMPT_INTERVAL = intPreferencesKey(name = "TOGGLE_REPORTS_PROMPT_INTERVAL")
+        val TOGGLE_REPORTS_MAX_PROMPT_COUNT = intPreferencesKey(name = "TOGGLE_REPORTS_MAX_PROMPT_COUNT")
+    }
+
+    private val toggleReportsRC: StateFlow<String> = store.data
```

Comment:
Is there a reason we load every property eagerly into a StateFlow (e.g. a performance issue you observed)? Have you considered making `canPrompt()` a `suspend fun` an reading from the data store only when needed?

Reply:
Honestly, I was mostly modeling after duck player because it was the most recent module with a datastore, but it is effectively a lot of overhead for probably marginal benefit so your suggestion makes sense to me.